File: app/src/main/java/crux/bphc/cms/models/NotificationSet.java
package crux.bphc.cms.models;

import io.realm.RealmModel;
import io.realm.annotations.PrimaryKey;
import io.realm.annotations.RealmClass;
import crux.bphc.cms.models.forum.Discussion;

/**
 * Created by harsu on 18-01-2017.
 */

@RealmClass
public class NotificationSet implements RealmModel {

    @PrimaryKey
    private int uniqueId;
    private int bundleId;
    private String notifSummary;
    private String notifTitle;
    private String notifContent;

    public NotificationSet() {
    }

    public NotificationSet(int uniqueId, int bundleId, String notifTitle, String notifContent, String notifSummary) {
        this.uniqueId = uniqueId;
        this.bundleId = bundleId;
        this.notifTitle = notifTitle;
        this.notifContent = notifContent;
        this.notifSummary = notifSummary;
    }

    /** Helper methods to create NotificationSet objects */
    public static NotificationSet createNotificationSet(Course course, CourseSection section, Module module) {
        return new NotificationSet(module.getId(), course.getId(), section.getName(), module.getName(), course.getShortname());
    }

    public static NotificationSet createNotificationSet(Course course, Module module, Discussion discussion) {
        return new NotificationSet(discussion.getId(), course.getId(), module.getName(), discussion.getMessage(), course.getShortname());
    }

    public static NotificationSet createNotificationSet(int courseID, String courseName, int modId, String moduleName) {
        return new NotificationSet(modId, courseID, "", moduleName, courseName);
    }

    public int getUniqueId() {
        return uniqueId;
    }

    public int getBundleID() {
        return bundleId;
    }


    public void setBundleID(int bundleID) {
        this.bundleId = bundleID;
    }

    public String getNotifSummary() {
        return notifSummary;
    }

    public void setNotifSummary(String notifSummary) {
        this.notifSummary = notifSummary;
    }

    public String getNotifTitle() {
        return notifTitle;
    }

    public void setNotifTitle(String notifTitle) {
        this.notifTitle = notifTitle;
    }

    public String getNotifContent() {
        return notifContent;
    }

    public void setNotifContent(String notifContent) {
        this.notifContent = notifContent;
    }

    public void setUniqueId(int uniqueId) {
        this.uniqueId = uniqueId;
    }

    public String getTitle() {
        return notifSummary;
    }

    public String getGroupKey() {
        return notifSummary;

    }

}

File: app/src/main/java/crux/bphc/cms/models/NotificationSet.kt
package crux.bphc.cms.models

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import crux.bphc.cms.activities.MainActivity
import crux.bphc.cms.app.Constants
import crux.bphc.cms.models.course.Course
import crux.bphc.cms.models.course.CourseSection
import crux.bphc.cms.models.course.Module
import crux.bphc.cms.models.forum.Discussion

/**
 * A data class for notification data
 *
 * @author Harshit Agarwal (18-Jan-2017)
 * @author Abhijeet Viswa
 */
data class NotificationSet(
        val uniqueId: Int,
        val bundleId: Int,
        val notifTitle: String,
        val notifContent: String,
        val groupKey: String?,
        val pendingIntent: PendingIntent?
) {
    companion object {
        /**
         * Helper method to create [NotificationSet] object for a new module
         */
        fun createNotificationSetForModule(context: Context, course: Course, section: CourseSection,
                                           module: Module)
                : NotificationSet {
            val intent = Intent(context, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_CLEAR_TASK or
                    Intent.FLAG_ACTIVITY_NEW_TASK
            intent.putExtra("courseId", course.id)
            intent.putExtra("modId", module.id)

            val pendingIntent = PendingIntent.getActivity(
                    context,
                    System.currentTimeMillis().toInt(),
                    intent,
                    PendingIntent.FLAG_UPDATE_CURRENT
            )
            return NotificationSet(module.id, course.id, section.name, module.name,
                    course.shortName, pendingIntent)
        }

        /**
         * Helper method to create [NotificationSet] object for a new discussion
         */
        fun createNotificationSetForDiscussion(context: Context, course: Course, module: Module,
                                               discussion: Discussion)
                : NotificationSet {
            val intent = Intent(context, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_CLEAR_TASK or
                    Intent.FLAG_ACTIVITY_NEW_TASK
            intent.putExtra("courseId", course.id)
            intent.putExtra("modId", module.id)
            intent.putExtra("forumId", module.instance)
            intent.putExtra("discussionId", discussion.id)

            val pendingIntent = PendingIntent.getActivity(
                    context,
                    System.currentTimeMillis().toInt(),
                    intent,
                    PendingIntent.FLAG_UPDATE_CURRENT
            )
            return NotificationSet(discussion.id, course.id, module.name, discussion.message,
                    course.shortName, pendingIntent)
        }

        fun createNotificationSetForSiteNews(context: Context, discussion: Discussion)
               : NotificationSet {
            val intent = Intent(context, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_CLEAR_TASK or
                    Intent.FLAG_ACTIVITY_NEW_TASK
            intent.putExtra("courseId", Constants.SITE_NEWS_COURSE_ID)
            intent.putExtra("forumId", 1)
            intent.putExtra("discussionId", discussion.id)

            val pendingIntent = PendingIntent.getActivity(
                    context,
                    System.currentTimeMillis().toInt(),
                    intent,
                    PendingIntent.FLAG_UPDATE_CURRENT
            )
            return NotificationSet(1, 0, "Site News", discussion.message,
                    "Site News", pendingIntent)
        }
    }
}

